kroxylicious-integration-tests/src/test/java/io/kroxylicious/proxy/encryption/EnvelopeEncryptionFilterIT.java

```diff
@@ -248,6 +257,78 @@ void produceAndConsumeEncryptedAndPlainTopicsAtSameTime(KafkaCluster cluster, To
         }
     }
 
+    /**
+     * Test that ensures that the record offsets returned by the broker are faithfully relayed to the client.
+     * @param cluster underlying kafka cluster
+     * @param compactedTopic topic configured for compaction.
+     * @param directConsumer consumer connected directly to the underlying kafka cluster
+     * @param testKmsFacade kms facade
+     * @throws Exception exception
+     */
+    @TestTemplate
+    @SuppressWarnings("java:S2925")
+    void offsetFidelity(@BrokerConfig(name = "log.cleaner.backoff.ms", value = "50") KafkaCluster cluster,
+                        @TopicConfig(name = "segment.ms", value = "125") @TopicConfig(name = "cleanup.policy", value = "compact") Topic compactedTopic,
+                        Consumer<String, String> directConsumer,
+                        TestKmsFacade<?, ?, ?> testKmsFacade)
+            throws Exception {
+        var testKekManager = testKmsFacade.getTestKekManager();
+        testKekManager.generateKek(compactedTopic.name());
+
+        var builder = proxy(cluster);
+
+        builder.addToFilters(buildEncryptionFilterDefinition(testKmsFacade));
+
+        try (var tester = kroxyliciousTester(builder);
+                var proxyProducer = tester.producer();
+                var proxyConsumer = tester.consumer()) {
+
+            proxyProducer.send(new ProducerRecord<>(compactedTopic.name(), "a", "a1"));
+            // Send two messages for key "b", the first will be eligible for compaction
+            proxyProducer.send(new ProducerRecord<>(compactedTopic.name(), "b", "b1"));
+            proxyProducer.send(new ProducerRecord<>(compactedTopic.name(), "b", "b2")).get(5, TimeUnit.SECONDS);
+
+            // Sleep for segment.ms so that the broker will begin a new segment when the next produce is received.
+            // The records in the first segment will become eligible for compaction.
+            Thread.sleep(125);
```

Comment:
Are segments rolled based on a background timer or on elapsed time between records? Just wondering if we need a longer delay to ensure the segment has actually been rolled. (ideally we would have some mechanism to detect the roll and await that, but we do not live in that sort of world)

Reply:
> Are segments rolled based on a background timer or on elapsed time between records? 

AFAIK it is actually the produce path that detects the need to roll the segment.  There's no background thread involvement. I talked about this in this thread. https://github.com/kroxylicious/kroxylicious/pull/850#discussion_r1441626972

Reply:
Yes, re-reading that comment did answer my question